### ISSUE TITLE
docker-maven-plugin expects a target/ directory to be created where the pom.xml is located 

### ISSUE THREAD

**Issue by TomHellier** [spotify/docker-maven-plugin#253]
## Description

When running docker-maven-plugin on a pom.xml which specifies modules, but has not plugins itself which would create a target directory, the build fails with the following error:
[ERROR] Failed to execute goal com.spotify:docker-maven-plugin:0.4.11:build (build-image) on project MyApp: Exception caught: basedir /Users/thellier/Documents/workspace/project/MyApp/target does not exist -> [Help 1].

If I create the target directory manually
`mkdir target` then the build will succeed.

## How to reproduce

If you create a pom.xml with the following contents
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <parent>
        <groupId>com.example</groupId>
        <artifactId>MyParentPom</artifactId>
        <version>0.9.4-SNAPSHOT</version>
    </parent> 

    <groupId>com.example.MyApp</groupId>
    <artifactId>MyApp</artifactId>
    <version>0.9.4-SNAPSHOT</version>
    <packaging>pom</packaging>

    <modules>
        <module>MySubApp1</module>
        <module>MySubApp2</module>
    </modules>

    <build>
	<plugins>
      <plugin>
        <groupId>com.spotify</groupId>
        <artifactId>docker-maven-plugin</artifactId>
        <version>0.4.11</version>
	<inherited>false</inherited>
        <configuration>
          <imageName>docker.example.com:5000/${user.name}/centos7-MyApp</imageName>
          <dockerDirectory>./MyApp</dockerDirectory>
          <serverId>my-docker-registry</serverId>
          <registryUrl>https://docker.example.com:5000/v2/</registryUrl>
          <imageTags>
            <imageTag>${project.version}</imageTag>
            <imageTag>latest</imageTag>
          </imageTags>
          <resources>
            <resource>
              <targetPath>/</targetPath>
              <directory>${project.build.directory}</directory>
              <include>${project.build.finalName}.jar</include>
            </resource>
          </resources>
        </configuration>
        <executions>
          <execution>
            <id>build-image</id>
            <phase>package</phase>
            <goals>
              <goal>build</goal>
            </goals>
          </execution>
          <execution>
            <id>push-image</id>
            <phase>deploy</phase>
            <goals>
              <goal>push</goal>
            </goals>
            <configuration>
              <imageName>${registryUrl}</imageName>
            </configuration>
          </execution>
        </executions>
      </plugin>
	</plugins>
	</build>
</project>

with simple modules, 

## What do you expect

mvn clean package
The docker image is produced successfully

## What happened instead

Build fails with the following error:
[ERROR] Failed to execute goal com.spotify:docker-maven-plugin:0.4.11:build (build-image) on project MyApp: Exception caught: basedir /Users/thellier/Documents/workspace/project/MyApp/target does not exist -> [Help 1].

## Software:

- docker version: Docker for Mac, Version 1.12.0-rc4-beta20 (build: 10404) (also fails on latest docker for centos (jenkins server)
- docker-maven-plugin version: 0.4.11
- maven version: 3.3.9

## Full backtrace

```text
[INFO] ------------------------------------------------------------------------
[INFO] Building MyApp 0.9.4-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.6.1:clean (default-clean) @ MyApp ---
[INFO] 
[INFO] --- docker-maven-plugin:0.4.11:build (build-image) @ MyApp ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] MyApp ................................. FAILURE [  0.086 s]
[INFO] MySubApp1 ................... SKIPPED
[INFO] MySubApp1 ................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:21 min
[INFO] Finished at: 2016-07-28T18:48:55+01:00
[INFO] Final Memory: 283M/1545M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.spotify:docker-maven-plugin:0.4.11:build (build-image) on project MyApp: Exception caught: basedir /Users/thellier/Documents/workspace//MyApp/target does not exist -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :MyApp
```

**Comment by TomHellier** [comment]
I may look into a pull request to fixing this, i think it would be as simple as creating the target directory, as the docker-maven-plugin is writing files in there, without actually guaranteeing the folder exists.

TagMojo.java:122 and

**Comment by TomHellier** [comment]
Ok, looks like the exception is thrown from the copyResources method in BuildMojo, just trying to test it and build the docker-maven-plugin so my mvnDebug can pick it up

**Comment by TomHellier** [comment]
I've raised a pull request to fix this bug. I'll be running my SNAPSHOT version of this build until this can get merged in and released, but it looks like Spotify have a nice healthy release schedule and this is a minor bug so hopefully will happen soonish. Thanks for creating a great plugin.

**Comment by mattnworb** [comment]
Hey @TomHellier thanks for the issue report and the PR. I think even with your PR, your pom.xml example will still have an issue - Maven is building the parent project (`MyApp`) before it builds the two submodules. You can see this in the order of the listing of projects in the `Reactor Summary` section. 

If you want to have the parent project build a Docker image containing the output of the two submodules, you'll have to tell Maven that the parent project depends on those two modules. Creating a dependency between parent and submodules might lead to a circular dependency (as the submodules depend on the parent Maven project also), so I'm not sure if this is even something that Maven can handle. One way to solve a possible circular dependency would be to have a third submodules which depends on the other two which contains the Dockerfile.

**Comment by TomHellier** [comment]
Hi, thanks for the reply, I'm quite new to maven so realise now that this was probably more of a problem with my configuration than this code. I just saw an exception that I could fix with a mkdir target and it built properly.

What I did was create a new sub project,

<modules>
    <module>MySubApp1</module>
    <module>MySubApp2</module>
    <module>MySubApp-docker-build</module>
</modules>

and that sub-module would be responsible for running the Dockerfile in its parents directory.

It was working for me because when I ran mvn clean install, it build the docker image with the sub-projects jars from the last time it ran, I noticed when I started to get my projects built by jenkins that actually my maven project was rubbish.
